### ISSUE TITLE
[Help] Design of turn-based but complex game

### ISSUE THREAD

**Issue by HenryTheCat** [gyscos/cursive#340]
Is there a reccomended, canonical way to use Cursive to build a turn-based (but otherwise rather complex) game?

The scenario I have in mind is the following: there is a `game` object which runs the whole game world, but it requires to receive inputs and to have its state displayed. The `minesweeper` example covers the scenario so far. Though, being the game rather complex, it requires to show incremental state changes without user intervention (basically, small animations).

To handle complex actions, `game` has a built-in event queue. When an input is received, events to be processed are added to a queue. Then, an `update` method has to be called to process them one-by-one until none are left. Every time an event is processed, I would like the screen to be refreshed and a certain time waited before processing the next event. That, I think, should suffice to create simple animations to represent the consequences of an action unfolding.

I guess I have to implement `View` for `Game`, and that is not an issue. What I cannot wrap my head around is how to display intermediate steps after user input. I tried to use callbacks and `cb_sink` but that didn't work. Setting fps does not help, and anyway it should not be necessary since the game is still fundamentally turn-based.

I believe related questions have been asked multiple times, but I could not find a straight answer. Somehow even the `progress` example didn't help me, even though that should be more or less what I need (maybe that is more of a true real-time example).

**Comment by gyscos** [comment]
Thanks for the inquiry!

In general, as soon as the complexity becomes non-trivial, I'd recommend moving game logic in its own thread, and communicating via channels with a `GameView` - especially if things should happen without user input. This will become useful when/if you need more threads to offload AI computation, network communication or other possibly blocking processes. Being tied to Cursive's timing (which is not guaranteed to be at a fixed refresh rate) would soon show limitations.

In this case, you would have your game logic running in a thread, with some communication channels:
* Input event channels fed by the GameView representing user actions
* GameState output channel to tell the GameView what to draw
* Cursive callback channels to trigger refreshes and other UI actions

Note that there are other ways to communicate the GameState - a simple `Mutex`, a double-buffering solution to prevent locks with a "last valid" state and a "currently being computed" state, some persistent data structure, ...

If you think this is overkill for your project (it's very possible!), you can use Cursive's timing mechanism by using `Cursive::set_fps` and running updates in the `View::layout` method, but be aware that `set_fps` sets an approximate _minimum_ refresh rate. Any user input will also trigger a refresh, so it's possible the actual time between calls to `layout` will be less than expected. Because of this, you would have to do some elapsed time check in there. Also, it means it'll spend a bit more resource refreshing the screen when nothing happens (unless you change the fps dynamically depending on the action, but it's easy to get it wrong if multiple things happen at the same time).

**Comment by gyscos** [comment]
If you only need timing for some animations, it's also possible to only react to input in `on_event`, and spawn a thread for the animation itself.

For example:

```rust
struct GameView {
    /// This represents the progress of the animation.
    animation_progress: Arc<AtomicUsize>,
}

impl cursive::view::View for GameView {
    // ...

    fn on_event(&mut self, event: Event) -> EventResult {

        // ...
        // Here we decided that a response to this action is to start an animation.

        // In this simple example it will be represented by a simple `usize`.
        self.animation_progress = Arc::new(AtomicUsize::new(0));
        let progress = Arc::clone(&self.animation_progress);

        EventResult::with_cb(|s| {
            let cb_sink = s.cb_sink().clone();
            thread::spawn(move || {
                // This thread will move the animation forward at a fixed pace.

                for _ in 0..10 {
                    // Advance by one step each second.
                    thread::sleep(Duration::from_secs(1));
                    progress.fetch_add(1, Ordering::SeqCst);
                }
            });
        })
    }

    // ...
}
```

This way, most of the timing can be kept simple and minimal (only reacting to inputs). The `draw` method would just check `self.animation_progress`.

**Comment by HenryTheCat** [comment]
This solution looks great, it should give me the power I need while being very clean and keeping everything simple!

Do you think these use-cases should be turned into examples, for future reference?

**Comment by gyscos** [comment]
There's never too many examples! If we can reduce it to something clear and easy to understand, it would most likely be very helpful!

**Comment by HenryTheCat** [comment]
I am yet not as advanced as communicating via threads, but I tried writing a custom main loop:

```Rust
fn main_loop(siv: &mut Cursive) {
    siv.refresh();
    const DELTA: Duration = Duration::from_millis(100);
    let mut last_frame = SystemTime::now();

    // The big event loop begins!
    while siv.is_running() {

        let current_time = SystemTime::now();
        if last_frame + DELTA < current_time {
            last_frame = current_time;

            update(siv);
            siv.refresh();
        }

        siv.step();
    }
}
```

Fundamentally I want all business as usual, but with a view that automatically updates its state and redraws itself every `DELTA` time. Such update is handled by the `update` function. Does this look like the right kind of thing to do, or are there obvious mistakes? In particular, I tried commenting out `siv.refresh()` but then the updates do not display.

In general this seems to work fine, even though the final effect is not perfectly smooth: it seems sometimes it stutters a bit (even when compiled in release mode). It might be due to something else than the main loop, though.
